### PR TITLE
base: Allow clearing queue(active recipe) fully for factory module (upd 12.10.20)

### PR DESCRIPTION
**Pull request by dd84ai**
a player command to clear active recipe from factory module

### CODE REVIEW

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 1]
```diff
@@ -930,6 +930,33 @@ namespace PlayerCommands
...
+					&& base->modules[index]->type != Module::TYPE_M_DOCKING
+					&& base->modules[index]->type != Module::TYPE_M_CLOAKDISRUPTOR))
+			{
+				PrintUserCmdText(client, L"ERR Not factory module");
+				return;
+			}
+
+			FactoryModule *mod = (FactoryModule*)base->modules[index];
+			if (mod->ClearRecipe())
+				PrintUserCmdText(client, L"OK Active recipe is scheduled for purge in a close time");
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:955]: a close time?

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 2]
```diff
@@ -961,9 +988,10 @@ namespace PlayerCommands
 		else
 		{
 			PrintUserCmdText(client, L"ERR Invalid parameters");
-			PrintUserCmdText(client, L"/base facmod [list|clear|add]");
+			PrintUserCmdText(client, L"/base facmod [list|clear|purge|add]");
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:991]: think we're going to need to do a better job of documenting clear vs. purge here

**Review by dsyalex — CHANGES_REQUESTED** [review]
please fix the PR title

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 1 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:955]: I in mean in 'a nearby time', fixed

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 2 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:991]: documented as
"|  clear <index> - clear queue, which starts from second item in building order for factory module at <index>");
"|  purge <index> - clear only active recipe, which is current first device in building process for factory module at <index>");

**Comment by dd84ai** [comment]
And let me guess, squash it.

**Review by dsyalex — DISMISSED** [review]
I think a `RECIPE` object is supposed to be a global thing across all factory mods across all bases, having a field of it used (and indeed blanking all fields) for scheduled purging on a per-factory-module basis feels dangerous.
I'm also unconvinced about the use of the term "purge". What about cancel?

**Comment by dd84ai** [comment]
> I think a `RECIPE` object is supposed to be a global thing across all factory mods across all bases, having a field of it used (and indeed blanking all fields) for scheduled purging on a per-factory-module basis feels dangerous.
> I'm also unconvinced about the use of the term "purge". What about cancel?

RECIPE, global thing? I can see it is being initialized as part of class BuildModule, FactoryModule classes. Which are used as objects to fill an array of modules in some player base instance. It is not looking to be used as global thing.

I used the way I used to clear the queue because... was afraid that 'Timer' function which works with the recipe too, could work at the wrong time during recipe clearing. I made a clearing in a way safe that is safe for multi-threading.

I chose 'purge' to give a hint, that you destroy consumed by recipe items at the same time. 'cancel' means items consumed by recipe would be returned in my opinion.

**Comment by dsyalex** [comment]
> > I think a `RECIPE` object is supposed to be a global thing across all factory mods across all bases, having a field of it used (and indeed blanking all fields) for scheduled purging on a per-factory-module basis feels dangerous.
> > I'm also unconvinced about the use of the term "purge". What about cancel?
> 
> RECIPE, global thing? I can see it is being initialized as part of class BuildModule, FactoryModule classes. Which are used as objects to fill an array of modules in some player base instance. It is not looking to be used as global thing.

```
Plugins/Public/base_plugin/Main.cpp:map<uint, RECIPE> recipes;
Plugins/Public/base_plugin/Main.cpp:                            RECIPE recipe;
Plugins/Public/base_plugin/Main.cpp:                            RECIPE recipe;
Plugins/Public/base_plugin/Main.cpp:            RECIPE recipe = recipes[recipe_name];
```

**Comment by dd84ai** [comment]
Additional problem to be corrected:
add canceling for buildmod too

**Comment by dd84ai** [comment]
> > > I think a `RECIPE` object is supposed to be a global thing across all factory mods across all bases, having a field of it used (and indeed blanking all fields) for scheduled purging on a per-factory-module basis feels dangerous.
> > > I'm also unconvinced about the use of the term "purge". What about cancel?
> > 
> > 
> > RECIPE, global thing? I can see it is being initialized as part of class BuildModule, FactoryModule classes. Which are used as objects to fill an array of modules in some player base instance. It is not looking to be used as global thing.
> 
> ```
> Plugins/Public/base_plugin/Main.cpp:map<uint, RECIPE> recipes;
> Plugins/Public/base_plugin/Main.cpp:                            RECIPE recipe;
> Plugins/Public/base_plugin/Main.cpp:                            RECIPE recipe;
> Plugins/Public/base_plugin/Main.cpp:            RECIPE recipe = recipes[recipe_name];
> ```

	string cfg_fileitems = string(szCurDir) + "\\flhook_plugins\\base_recipe_items.cfg";
	if (ini.open(cfg_filemodules.c_str(), false))
	{
		while (ini.read_header())
		{
			if (ini.is_header("recipe"))
			{
				RECIPE recipe;
				while (ini.read_value())
				{
					if (ini.is_value("nickname"))
					{
						recipe.nickname = CreateID(ini.get_value_string(0));
					}
					else if (ini.is_value("produced_item"))
					{
						recipe.produced_item = CreateID(ini.get_value_string(0));
					}
					else if (ini.is_value("infotext"))
					{
						recipe.infotext = stows(ini.get_value_string());
					}
					else if (ini.is_value("cooking_rate"))
					{
						recipe.cooking_rate = ini.get_value_int(0);
					}
					else if (ini.is_value("consumed"))
					{
						recipe.consumed_items[CreateID(ini.get_value_string(0))] = ini.get_value_int(1);
					}
					else if (ini.is_value("reqlevel"))
					{
						recipe.reqlevel = ini.get_value_int(0);
					}
				}
				recipes[recipe.nickname] = recipe;
			}
		}
		ini.close();
	}

According to those strings, this variable just holds loaded possible list of recipes. (and nowhere related to queue)

**Comment by dd84ai** [comment]
> I'm also unconvinced about the use of the term "purge". What about cancel?

I don't mind. lets do it 'cancel'

**Comment by dd84ai** [comment]
> Additional problem to be corrected:
> add canceling for buildmod too

Update: no need to cancel buildmod modules, it is already possible to destroy them (even before the current change)

**Comment by dd84ai** [comment]
Optimized work. Removed all unnecessary steps. Clearing active recipe is much.... simpler now.

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 3]
```diff
@@ -1005,9 +1030,10 @@ namespace PlayerCommands
 		else
 		{
 			PrintUserCmdText(client, L"ERR Invalid parameters");
-			PrintUserCmdText(client, L"/base facmod [list|clear|add]");
+			PrintUserCmdText(client, L"/base facmod [list|clear|cancel|add]");
 			PrintUserCmdText(client, L"|  list - show factory modules and build status");
-			PrintUserCmdText(client, L"|  clear <index> - clear build queue for factory module at <index>");
+			PrintUserCmdText(client, L"|  clear <index> - clear queue, which starts from second item in building order for factory module at <index>");
+			PrintUserCmdText(client, L"|  cancel <index> - clear only active recipe, which is current first device in building process for factory module at <index>");
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:1036]:
```suggestion
			PrintUserCmdText(client, L"|  cancel <index> - clear only active recipe, which is the first item in the building queue for the factory module at <index>");
			
```

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 4]
```diff
@@ -1005,9 +1030,10 @@ namespace PlayerCommands
 		else
 		{
 			PrintUserCmdText(client, L"ERR Invalid parameters");
-			PrintUserCmdText(client, L"/base facmod [list|clear|add]");
+			PrintUserCmdText(client, L"/base facmod [list|clear|cancel|add]");
 			PrintUserCmdText(client, L"|  list - show factory modules and build status");
-			PrintUserCmdText(client, L"|  clear <index> - clear build queue for factory module at <index>");
+			PrintUserCmdText(client, L"|  clear <index> - clear queue, which starts from second item in building order for factory module at <index>");
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:1035]:
```suggestion
			PrintUserCmdText(client, L"|  clear <index> - clear queue, which starts from the second item in the building queue for the factory module at <index>");

```

**Review by dsyalex — COMMENTED** [review]
doc wording tweaks but looking good

**Pull request merged**